pallets/proofs-dealer/src/utils.rs

```diff
@@ -620,6 +651,59 @@ where
 
         challenges
     }
+
+    /// Trim the storage that holds the Providers that submitted valid proofs in the last blocks until there's
+    /// `TargetBlocksOfProofsStorage` blocks left (or until the remaining weight allows it).
+    ///
+    /// This function is called in the `on_idle` hook, which means it's only called when the block has
+    /// unused weight.
+    ///
+    /// It removes the oldest block from the storage that holds the providers that submitted valid proofs
+    /// in the last blocks as many times as the remaining weight allows it, but at most until the storage
+    /// has `TargetBlocksOfProofsStorage` blocks left.
+    pub fn do_trim_valid_proof_submitters_last_blocks(
+        _n: BlockNumberFor<T>,
+        usable_weight: Weight,
+    ) -> Weight {
+        // Initialize the weight used by this function.
+        let mut used_weight = Weight::zero();
+
+        // Since we are using a double map and we don't know how many Providers are going to exist in each block, we delete
+        // not blocks but chunks of Providers. This means blocks older than "TargetBlocksOfProofsStorage" ago are in an undefined
+        // state and should not be used. This is a trade-off to avoid iterating over all Providers in each block, which could be
+        // computationally expensive.
+        let weight_to_remove_block = T::DbWeight::get().reads_writes(1, 1); // TODO: Benchmark this logic to get the actual weight.
+        let removable_blocks = usable_weight.checked_div_per_component(&weight_to_remove_block);
+
+        // If there is enough weight to remove blocks, try to remove them.
+        if let Some(removable_blocks) = removable_blocks {
+            // Check to see if our limiting factor for removing old blocks is weight remaining or the target amount of blocks to keep in storage.
+            let blocks_in_storage = ValidProofSubmittersLastTicks::<T>::count();
+            let target_blocks = TargetTicksStorageOfSubmittersFor::<T>::get();
+            let removable_blocks: u32 = removable_blocks
+                .try_into()
+                .unwrap_or(u32::MAX)
+                .min(blocks_in_storage.saturating_sub(target_blocks));
+
+            // Remove all the blocks that we can.
+            for _ in 0..removable_blocks {
+                // Get the oldest block.
+                let mut available_blocks = ValidProofSubmittersLastTicks::<T>::iter_keys()
+                    .collect::<Vec<BlockNumberFor<T>>>();
+                available_blocks.sort();
+                let oldest_block = available_blocks.first().expect("We checked before how many blocks are available and at least one should be there.");
```

Comment:
First: don't use expect here. Second, this can actually be optimised by having a variable keeping track of the last deleted tick. It's a small price to pay for avoiding collecting and ordering a vec.